Fix visual stuttering on WindowsClassic version

dobragab opened the pull request:
While the ball moves, it visibly stutters, as if there was an FPS drop.

Turns out that Windows timer resolution, more specifically `Sleep()` is the root cause, its precision is too high.

Set the timer precision to 1ms and sleep for 1ms only. This still doesn't cause much CPU usage but provides a significantly better visual experience.

k4zmu2a commented on SpaceCadetPinball/winmain.cpp:
```diff
@@ -311,7 +319,7 @@ int winmain::WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPSTR lpCmdLi
 				if (curTime == then)
 				{
-					Sleep(8u);
+					Sleep(1u);
```
```suggestion
					Sleep(8u);
```

k4zmu2a commented:
timeBeginPeriod(1) is a valid addition, without it game cannot reach 120 UPS.
Sleep should remain at 8 ms.

dobragab commented:
Hi,

There are multiple things wrong with `Sleep(8)`:

* First, it means that (given 1 ms timer precision) it will actually sleep for either 8 or 9 milliseconds. 9 is definitely too much for 120 FPS, as 120 FPS means 8.3 ms frame time.
* Second, there are monitors with higher refresh rate, like mine at 144 Hz (which is probably the highest available on the consumer market). 144 FPS means 6.94 ms frame time. So, to satisfy 144 Hz monitors, `Sleep(5)` should be enough, which might sleep for 6 ms, and there's 0.94 ms available to perform actual rendering.
* Third, the average FPS was already pretty high, the problem is with frame time spikes. Sometimes the frame time jumps above 7 ms even when it's `Sleep(1)`. 

I added a printf to `pb::frame` to print frame time when it reaches 7 ms, and then tested against different sleep values:
* 1 ms gives almost perfect results, it sometimes skips frames only when I'm hammering both bumpers. 
* 2 ms often skips frames when I'm hammering both bumpers, and skips really rarely if I'm using one bumper, usually when the ball's up and noisy.
* 3-5 ms often skips when I'm using one bumper, i.e. during normal gameplay too.

So I'd say 2 ms is the good compromise, it doesn't hurt gameplay experience, not even on 144 Hz monitors, and the CPU usage is basically nothing, it's under 1% on my machine. Yes, I tested a Release build.

I've also done my homework: some analysis about the frame skips, with Visual Studio profiler (in release mode). The repro is, just hammer both bumper keys all the time. The symptom is: you get approximately one spike for each time you press both bumper keys at the same time.

It looks like that it stresses `pb::keydown` and `pb::keyup` functions, both of which spends most of its time in `TFlipper::Message`, more specifically `loader::play_sound`. While I'm not sure that this causes the spike, given the symptoms, my guess is that the root cause is the keyup and keydown sounds overlapping a little, causing many wavs to be played. And if that's the case, I'm pretty sure that we can't really get rid of that without deviating from the original behavior.

I also found some performance improvements which also slightly reduce frame skipping:
* More optimization flags in Release, I enabled `/Ob2` and `/Ot`.
* The program spends approx. 40% of its CPU time in `WaveMix::FreePlayedBlocks` which isn't reasonable. I changed winmain to call it less frequently, i.e. not at every frame. My measurements above were performed with this change.

I'm happy to share these improvements, I'm unsure whether it belongs to this PR or a new one.

k4zmu2a commented:
That’s all well and good, however I have no plans to change main loop in WindowsClassic version of the game.
Subject of main loop rebalance is handled in v2.
WindowsClassic version is meant to stay ‘true to the original’ TM.

I am ok with fixing glaring issues, like no timeBeginPeriod when the game clearly needs it.
I don’t consider slightly unstable main loop a big issue.

It is ok to add more optimizations in release build if code can handle it.
I think there is no need to throttle Sound::Idle at 120UPS.

dobragab commented:
I don't really want to fix main loop rebalance either, it's fine as is.

I tested it a bit deeper, and even `Sleep(4)` has the visual stuttering on my 144 Hz monitor, just not as bad as in the original, so I'd consider this as a glaring issue. `Sleep(2)` is  fine though, I'll update my PR to that, and open a new one for small perf optimizations.

P.S. Pretty funny that a [MS dev stated](https://devblogs.microsoft.com/oldnewthing/20051201-09/?p=33133) that his proudest fix was (not just "caused", it literally _was_) the very bug this PR attempts to solve. It's even stated on the game's [Wikipedia article](https://en.wikipedia.org/wiki/Full_Tilt!_Pinball#3D_Pinball_for_Windows_%E2%80%93_Space_Cadet).

k4zmu2a commented:
I am not merging this, but I will cherry pick timeBegin.

Buts let’s dig deeper anyway.
First, what is your definition of “visual stuttering”?

Second, there is a known problem with GDI - it introduces tearing if you draw partial updates faster than VSync.
The ball is most affected by this, especially when moving fast.
I introduced a hack to combat this, hidden under alternative rendering option.

Now, you want to render at 2ms ~500 FPS (or even at 1ms), this will greatly accentuate the GDI problem.
One way to solve GDI problem is to match FPS to VSync, for example with options like in v2.
I don’t plan to do that.

Rendering system caveats:
FPS are tied to UPS, no way to decouple while keeping retained render.
The one alternative (used in GDI hack), full screen blit consumes a lot of CPU, especially when scaling.

I find that UPS = 2xFPS works best for game physics.
That is why I settled on vanilla 120 UPS/FPS and 60 FPS 120 UPS in redraw mode.
High refresh rate monitors are not a big concern for v1.

k4zmu2a commented:
Frame times were really out of hand, I did not expect that magnitude of error.
Added loop stabilization, it should be better now.
Stable frame times somewhat improved GDI tearing, maybe by making more repeatable, I am not sure.

Also “cherry picked” timeBegin.

dobragab commented:
(Sorry for the rather large delays, as well as for the other PRs.)

I meant by "visual stuttering" is basically microlags, when updates for some frames visibly skipped. 

For a reason that I don't fully understand, I didn't notice tearing *at all*, though my primary monitor is 60 Hz, FreeSync is turned off, and I tested my changes on both of my monitors. Yesterday I did try my build on two other PCs, and I saw the horrible tearing, so now I understand your concerns much better.

I checked your changes, and they make things *a lot* better. My only issue currently is that target frame rate is too low for 144 Hz. I can decline this PR and open a new one to support high frame rate, it's fairly easy to calculate desired frame times both for default and alternative rendering (and also deal with fractions properly). 

(BTW I would have slightly preferred actual cherry picking, to preserve author history, but I don't care much.)

Is it OK for you to add a new registry key e.g. `TargetFrameRate` that controls this, and default to 60 if not found?

k4zmu2a commented:
Adding target frame rate as registry option for advanced users is fine.
I just don’t want it in the menus.

To achieve 144 UPS exactly you will need float frame time, I will add that later.
The closest you can get with int without period is 1000/7.0 UPS.

dobragab commented:
I added a registry option for Target FPS. What do you think? Shall I resolve the conflict with your latest changes?

k4zmu2a commented:
New option is wrong in several places.
Option should be TargetUPS, clamped after reading at [60,360], default 120.
Main loop should use it as is – TargetFrameTime = 1000 / TargetUPS.
Alternative render hack should have FPS = TargetUPS / 2, but no less than 60.
Default value of new option should be written to registry in options::uninit.

Naturally, you should write against the latest code you want anything merged)

dobragab commented:
> Option should be TargetUPS, clamped after reading at [60,360], default 120.
> Alternative render hack should have FPS = TargetUPS / 2, but no less than 60.

I'd rather still call it TargetFPS, as it isn't trivial that FPS = TargetUPS / 2 when you set the registry entry.

k4zmu2a commented:
I don’t get it, are you talking about alternative rendering hack?
In regular display mode FPS == UPS.
This option primary use is to match UPS to display refresh rate in regular draw mode.

dobragab commented:
Okay, I'll adjust my changes accordingly.

dobragab commented:
Updated. Defaulted to 125 UPS, it seems slightly smoother on a 60 Hz display.

k4zmu2a commented:
It is closer now, but still not quite to my liking.
Issue list:
•	Winmain: don’t use int, don’t limit TargetFrameTime
•	Options: TargetUps should be clamped [60,360], default should be 120, option is not saved
•	Pb: targetTime = 1000 / (TargetUps / 2.0f), TargetUps/2.0f should be no less than 60, do not alter frameCounter

I you want, I can merge this as is and fix it later.

dobragab commented:
Fixed all three, they were pretty straightforward.

k4zmu2a commented on SpaceCadetPinball/pb.cpp:
```diff
@@ -270,15 +270,21 @@ void pb::frame(float dtMilliSec)
+			targetFps = max(targetFps, 60); // at least 60
+
+			// Frame time at 60 FPS = 16.(6) ms = (16 + 17 + 17) / 3
+			auto targetTime = (frameCounter + 1) * 1000 / targetFps - frameCounter * 1000 / targetFps;
```
targetTime should be: auto targetTime = 1000 / targetFps; 
targetFps should be: auto targetFps = options::Options.TargetUps / 2.0f;

k4zmu2a commented on SpaceCadetPinball/pb.cpp:
```diff
@@ -270,15 +270,21 @@ void pb::frame(float dtMilliSec)
 				frameCounter++;
 			}
+
+			frameCounter %= targetFps;
```
This is not needed.

k4zmu2a commented on SpaceCadetPinball/winmain.cpp:
```diff
@@ -225,9 +223,14 @@ int winmain::WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPSTR lpCmdLi
 
 	DWORD someTimeCounter = 300u, prevTime = 0u, frameStart = timeGetTime();
-	float sleepRemainder = 0, frameDuration = TargetFrameTime;
+	float sleepRemainder = 0, frameDuration = (float)TargetFrameTime;
```
```suggestion
	float sleepRemainder = 0, frameDuration = TargetFrameTime;
```

k4zmu2a commented on SpaceCadetPinball/winmain.cpp:
```diff
@@ -326,7 +329,7 @@ int winmain::WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPSTR lpCmdLi
 
 				sleepRemainder = max(min((int)(frameEnd - updateEnd) - sleepDuration, TargetFrameTime), -TargetFrameTime);
-				frameDuration = min((frameEnd - frameStart), TargetFrameTime * 2);
+				frameDuration = min(frameEnd - frameStart, TargetFrameTime * 2);
```
```suggestion
				frameDuration = min((frameEnd - frameStart), TargetFrameTime * 2);
```

k4zmu2a commented:
Sorry, but it is still not quite there yet.
I will try a more direct highlighting approach with comments.
The comments address issues and remove unnecessary changes.

k4zmu2a commented:
After a long journey, it is finally done)
Thank you for sticking with my seemingly unreasonable demands.
